File: pine-java/src/main/java/page/liam/pine/operators/TransformBenchSleep.java
package page.liam.pine.operators;

import page.liam.pine.AbstractOperator;
import page.liam.pine.CancellationToken;
import page.liam.pine.ConcurrentSafe;
import page.liam.pine.OperatorInput;
import page.liam.pine.OperatorOutput;
import page.liam.pine.OperatorParams;

public class TransformBenchSleep extends AbstractOperator implements ConcurrentSafe {
    private int delayMs = 5;

    @Override
    public void init(OperatorParams params) {
        Object v = params.get("delay_ms");
        if (v instanceof Number) {
            delayMs = ((Number) v).intValue();
        }
    }

    @Override
    public void execute(CancellationToken token, OperatorInput input, OperatorOutput output) {
        try {
            Thread.sleep(delayMs);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        int n = input.itemCount();
        for (int i = 0; i < n; i++) {
            output.setItem(i, "_bench_slept", true);
        }
    }
}

File: pine-cpp/operators/transform/transform_bench_sleep.cpp
#include "pine/operator.hpp"

#include <chrono>
#include <thread>

#include "operators/_helpers.hpp"

namespace pine {

class TransformBenchSleepOp : public Operator, public ConcurrentSafe {
 public:
  void init(const OperatorConfig& cfg) override {
    auto it = cfg.params.as_object().find("delay_ms");
    if (it != cfg.params.as_object().end() && it->second.is_number()) {
      delay_ms_ = static_cast<int>(it->second.as_number());
    }
  }

  void execute(const OperatorInput& input, OperatorOutput& out) override {
    std::this_thread::sleep_for(std::chrono::milliseconds(delay_ms_));
    for (std::size_t i = 0; i < input.item_count(); ++i) {
      out.set_item(static_cast<int>(i), "_bench_slept", Variant(true));
    }
  }

 private:
  int delay_ms_ = 5;
};

static const OperatorSchema k_transform_bench_sleep_schema{
    .name = "transform_bench_sleep",
    .type = OpType::Transform,
    .description =
        "Benchmark-only I/O-simulating operator. Sleeps for delay_ms per invocation. Not available in "
        "pine-python.",
    .params =
        {
            {"delay_ms",
             {.type = "int",
              .required = false,
              .default_value = Variant(5.0),
              .description = "Milliseconds to sleep per operator invocation."}},
        },
    .metadata = {.common_input = "", .common_output = "", .item_input = "", .item_output = ""},
};
PINE_REGISTER_OPERATOR_T(TransformBenchSleepOp, k_transform_bench_sleep_schema)

}  // namespace pine
